### PR TITLE
Java: Updated JNI

### LINKED ISSUE #1894: Update JNI on java bindings
https://github.com/jni-rs/jni-rs/releases/tag/v0.21.0

### PR DESCRIPTION
**Pull request by kwek20**
# Description of change

Updated jni to new version

## Links to any relevant issues

Fixed https://github.com/iotaledger/wallet.rs/issues/1894

### CODE REVIEW

File: wallet/bindings/java/lib/native/src/lib.rs [hunk 1]
```diff
@@ -49,7 +49,7 @@ macro_rules! jni_err_assert {
 // it throws an exception and returns the value of r. Otherwise returns a String.
 macro_rules! string_from_jni {
     ($env:ident, $x:ident, $r:expr ) => {{
-        let string = $env.get_string($x);
+        let string = $env.get_string_unchecked(&$x);
```
Comment (thibault-martinez) [inline wallet/bindings/java/lib/native/src/lib.rs:52]: Is this alright? I don't remember having to do this when I updated JNI in the client 🤔

File: wallet/bindings/java/lib/native/src/lib.rs [hunk 2]
```diff
@@ -86,8 +90,8 @@ pub extern "system" fn Java_org_iota_api_NativeApi_initLogger(env: JNIEnv, _clas
 
 // This keeps rust from "mangling" the name and making it unique for this crate.
 #[no_mangle]
-pub extern "system" fn Java_org_iota_api_NativeApi_createMessageHandler(
-    env: JNIEnv,
+pub unsafe extern "system" fn Java_org_iota_api_NativeApi_createMessageHandler(
```
Comment (thibault-martinez) [inline wallet/bindings/java/lib/native/src/lib.rs:93]: Why is this needed now ?

File: wallet/bindings/java/lib/native/src/lib.rs [hunk 3]
```diff
@@ -247,16 +254,15 @@ unsafe fn event_handle(clazz: GlobalRef, callback_ref: GlobalRef, event: Event)
...
 
     // Get a ref back to the callback we call on the java side
-    let cb = JValue::Object(callback_ref.as_obj()).to_jni();
-    let clazz = JClass::from_raw(*clazz.as_obj());
+    let cb = JValue::Object(callback_ref.as_obj()).as_jni();
 
     // Call NativeApi, METHOD_CACHE assumed initialised if we received a VM
     let res = env.call_static_method_unchecked(
-        clazz,
+        &clazz,
```
Comment (thibault-martinez) [inline wallet/bindings/java/lib/native/src/lib.rs:265]: Just curious, is `clazz` a common thing in Java?

File: wallet/bindings/java/lib/native/src/lib.rs (hunk 2 above)
Comment (kwek20) [reply wallet/bindings/java/lib/native/src/lib.rs:93]: unsafe is needed for env-throw_exception()

File: wallet/bindings/java/lib/native/src/lib.rs (hunk 3 above)
Comment (kwek20) [reply wallet/bindings/java/lib/native/src/lib.rs:265]: yes, as class is a keyword, you write clazz.i suppose it wasnt needed as its rust, but i was used to it.

File: wallet/bindings/java/lib/native/src/lib.rs (hunk 1 above)
Comment (kwek20) [reply wallet/bindings/java/lib/native/src/lib.rs:52]: yea, its slightly faster then the get_string version as it.. doesnt check the type is a java.lang.String. We check all this on java side

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]
